text/000-sfc-and-template-import-primitives.md

```diff
@@ -0,0 +1,875 @@
+
+Component subclasses inherit their template from the nearest superclass that
+has had a template set via `setComponentTemplate`. If no superclass has an
+associated template, Ember will fall back to the current template resolution
+rules.
+
+Templates set via `setComponentTemplate`, including those set on a
+superclass, take precedence over the `layout` property, container lookup, or
+any other existing mechanism of template resolution. For example, a template
+set on a superclass via `setComponentTemplate` would take precedence over a
+subclass's `layout` property (if it had one). However, a template set via
+`setComponentTemplate` on the subclass would take precedence over the
+superclass.
+
+### Template Scope
+
+In this context, _scope_ refers to the set of names that can be used to refer
+to values, helpers, and other components from within your template. For
+example, when you type `{{@firstName}}`, `{{this.count}}`, or `<UserAvatar
+/>`, how does Ember know what each of these refers to? The answer to that is
+determined by the template's scope.
+
+Today, the names available inside a particular template are determined by a
+few different rules:
+
+1. `this` always refers to the component instance (if the template has a
+   backing class).
+2. Arguments (like `@firstName`) are placed into scope when you invoke the
+   component.
+3. Block parameters (like `item` in `{{#each @items as |item|}}`) are in
+   scope, but only inside the block.
+4. Anything else, like components and helper names, must be looked up
+   dynamically through Ember's container. This lookup process is somewhat
+   complex and not always as fast as we'd like.
+
+Because the rules for how components and helpers are looked up are implicit
+and can be confusing for new learners, we'd like to explore alternate APIs
+that allow users to explicitly import values and refer to them from
+templates.
+
+This RFC proposes that templates embedded via the `createEmbeddableTemplate`
+function resolve unknown identifiers from the ambient JavaScript scope,
+rather than performing dynamic resolution through the container.
+
+To do this, `createEmbeddableTemplate` emits JavaScript code that contains
+references to the same identifiers as those used in the template:
+
+```js
+import { t } from '../helpers/t';
+import User from './user';
+import { __template__ } from '@ember/template-compiler';
+
+const template$1 = __template__({
+  source: '{{t "Hello!"}} <User @user={{this.currentUser}} />',
+  mode: 1,
+  scope: [User, t]
+});   /*  ^------- actual JavaScript identifier */
+```
+
+When the embedded template is later compiled into the wire format, these references
+are retained:
+
+```js
+import { t } from '../helpers/t';
+import User from './user';
+import { templateFactory } from '@ember/template-factory';
+
+const template$1 = templateFactory({
+  "id": "ANJ73B7b",
+  "block": "{\"statements\":[\"...\"]}",
+  "scope": () => [User, t]
```

Comment:
are there memory implications of maintaining explicit references to these values? I don't know, I only mean this is explicitly different from ember's current architecture. consider a page with hundreds or thousands of the same component (like a table cell, for example). will that page have noticeably different memory constraints on this RFC than it does in ember today?